core/src/main/java/com/linecorp/armeria/common/DefaultHttpHeaders.java

```diff
@@ -186,19 +189,23 @@ public HttpHeaders status(HttpStatus status) {
     @Nullable
     @Override
     public MediaType contentType() {
-        final MediaType contentType = this.contentType;
-        if (contentType != null) {
-            return contentType;
-        }
-
         final String contentTypeString = get(HttpHeaderNames.CONTENT_TYPE);
         if (contentTypeString == null) {
             return null;
         }
 
+        if (contentType != null) {
+            if (contentTypeString.equals(this.contentTypeString) ||
+                contentTypeString.equals(contentType.toString())) {
+                return this.contentType;
+            }
+        }
+
         try {
-            this.contentType = MediaType.parse(contentTypeString);
-            return this.contentType;
+            contentType = MediaType.parse(contentTypeString);
```

Comment:
nit: How about setting `contentType` and `this. contentTypeString` to `null` before parsing?